km61/Chala_Hanna/laboratory1.sql

```diff
@@ -1,2 +1,120 @@
 -- LABORATORY WORK 1
 -- BY Chala_Hanna
+/*---------------------------------------------------------------------------
+1.Створити схему даних з назвою – прізвище студента, у якій користувач може: 
+змінювати структуру таблиць
+4 бали
+
+---------------------------------------------------------------------------*/
+--Код відповідь:
+CREATE USER CHALA IDENTIFIED BY 123
+DEFAULT TABLESPACE = "USERS"
+TEMPORARY TABLESPACE = "TEMP"
+ALTER USER CHALA QUOTA '100M' TO USER
+GRANT "CONNECT" TO CHALA;
+GRANT ALTER ANY TABLE TO CHALA;
+
+
+
+
+
+
+
+
+
+
+/*---------------------------------------------------------------------------
+2. Створити таблиці, у яких визначити поля та типи. Головні та зовнішні ключі 
+створювати окремо від таблиць використовуючи команди ALTER TABLE. 
+Студент має залікову, що містить записи про дисципліни.
+4 бали
+
+---------------------------------------------------------------------------*/
+--Код відповідь:
+
+CREATE TABLE STUDENT (st_name VARCHAR(30) NOT NULL);
+CREATE TABLE ZALIK(lesson VARCHAR(10) NOT NULL,
+mark NUMBER(8,2) NULL);
+ALTER TABLE STUDENT
+ADD CONSTRAINT name_pk PRIMARY KEY (st_name);
+ALTER TABLE ZALIK 
+ADD CONSTRAINT lesson_pk PRIMARY KEY (lesson);
+CREATE TABLE ZALIK_STUDENT
+(st_name_fk VARCHAR(30) NOT NULL,
+lesson_fk VARCHAR(10) NOT NULL,
+mark NUMBER(8,2) NULL);
+ALTER TABLE ZALIK_STUDENT
+ADD CONSTRAINT st_name_pk PRIMARY KEY (st_name_fk);
+ALTER TABLE ZALIK_STUDENT 
+ADD CONSTRAINT z_lesson_pk PRIMARY KEY (lesson_fk);
+ALTER TABLE ZALIK_STUDENT
+ADD CONSTRAINT studname_fk FOREIGN KEY (st_name_fk) REFERENCE STUDENT (st_name);
+ALTER TABLE ZALIK_STUDENT 
+ADD CONSTRAINT zlesson_fk FOREIGN KEY (lesson_fk) REFERENCE ZALIK (lesson);
```

Comment:
`+ADD CONSTRAINT studname_fk FOREIGN KEY (st_name_fk) REFERENCE STUDENT (st_name);
+ALTER TABLE ZALIK_STUDENT 
+ADD CONSTRAINT zlesson_fk FOREIGN KEY (lesson_fk) REFERENCE ZALIK (lesson);`

Misspelling of the keyword. It should be referenceS instead of reference. Fix it in both cases.

Reply:
CREATE TABLE STUDENT1234 (ST_NAME VARCHAR(30) NOT NULL,
ST_GROUP varchar(5) NULL,
ST_PASPORT INT NULL,
ST_FACULTY VARCHAR(20) NOT NULL);
CREATE TABLE ZALIK1234(ED_LEVEL VARCHAR(10) NOT NULL,
NUMB INT NULL,
YEAR INT NOT NULL,
STUDY_FORM VARCHAR(30) NOT NULL);
CREATE TABLE UNIVERSITY1234(SUBJECT VARCHAR(10) NOT NULL,
SEMESTR NUMBER(8,2) NULL,
SUBJ_ID INT NOT NULL,
TEACHER VARCHAR(30) NOT NULL);
ALTER TABLE STUDENT1234
ADD CONSTRAINT PK1 PRIMARY KEY (ST_PASPORT);
ALTER TABLE ZALIK1234 
ADD CONSTRAINT PK2 PRIMARY KEY (NUMB);
ALTER TABLE UNIVERSITY1234
ADD CONSTRAINT PK3 PRIMARY KEY (SUBJ_ID);
ALTER TABLE ZALIK1234
ADD CONSTRAINT FK1 FOREIGN KEY (NUMB) REFERENCES STUDENT1234 (ST_PASPORT);
CREATE TABLE ZALIK_STUDENT1234
(NUMB_fk INT NOT NULL,
SUBJ_ID_fk INT NOT NULL,
MARK VARCHAR(20) NOT NULL);


ALTER TABLE ZALIK_STUDENT1234
ADD CONSTRAINT PK4 PRIMARY KEY (NUMB_fk, SUBJ_ID_fk);

ALTER TABLE ZALIK_STUDENT1234
ADD CONSTRAINT FK2 FOREIGN KEY (NUMB_fk) REFERENCES ZALIK1234 (NUMB);
ALTER TABLE ZALIK_STUDENT1234 
ADD CONSTRAINT FK3 FOREIGN KEY (SUBJ_ID_fk) REFERENCES UNIVERSITY1234 (SUBJ_ID);

Reply:
![capture](https://user-images.githubusercontent.com/22836311/38839208-3a2880b4-41e2-11e8-93d0-31c7bf962ab6.PNG)
![capture1](https://user-images.githubusercontent.com/22836311/38839209-3a45d678-41e2-11e8-9bcf-4d52bd9a62d1.PNG)

Reply:
INSERT INTO STUDENT1234(ST_NAME, ST_GROUP, ST_PASPORT, ST_FACULTY)
VALUES('ANIA', 'KM-62', '10001', 'FPM');
INSERT INTO STUDENT1234(ST_NAME, ST_GROUP, ST_PASPORT, ST_FACULTY)
VALUES('ANN', 'DP-61', '10002', 'FEL');
INSERT INTO STUDENT1234(ST_NAME, ST_GROUP, ST_PASPORT, ST_FACULTY)
VALUES('HANNA', 'KM-61', '10003', 'FPM');
INSERT INTO STUDENT1234(ST_NAME, ST_GROUP, ST_PASPORT, ST_FACULTY)
VALUES('N', 'FI-63', '10004', 'FTI');
INSERT INTO ZALIK1234(ED_LEVEL, NUMB, YEAR, STUDY_FORM)
VALUES('BACH', '10001', '2016', 'DN');
INSERT INTO ZALIK1234(ED_LEVEL, NUMB, YEAR, STUDY_FORM)
VALUES('BACH', '10002', '2015', 'DN');
INSERT INTO ZALIK1234(ED_LEVEL, NUMB, YEAR, STUDY_FORM)
VALUES('BACH', '10003', '2016', 'DN');
INSERT INTO ZALIK1234(ED_LEVEL, NUMB, YEAR, STUDY_FORM)
VALUES('MAG', '10004', '1012', 'ZAOCH');
INSERT INTO UNIVERSITY1234(SUBJECT, SEMESTR, SUBJ_ID, TEACHER)
VALUES('MATHS', '1', '111', 'A');
INSERT INTO UNIVERSITY1234(SUBJECT, SEMESTR, SUBJ_ID, TEACHER)
VALUES('MATHS', '1', '222', 'B');
INSERT INTO UNIVERSITY1234(SUBJECT, SEMESTR, SUBJ_ID, TEACHER)
VALUES('DB', '1', '333', 'C');
INSERT INTO UNIVERSITY1234(SUBJECT, SEMESTR, SUBJ_ID, TEACHER)
VALUES('DIF', '1', '444', 'A');
INSERT INTO ZALIK_STUDENT1234(NUMB_FK, SUBJ_ID_FK, MARK)
VALUES('10001', '111', '5');
INSERT INTO ZALIK_STUDENT1234(NUMB_FK, SUBJ_ID_FK, MARK)
VALUES('10001', '333', '4');
INSERT INTO ZALIK_STUDENT1234(NUMB_FK, SUBJ_ID_FK, MARK)
VALUES('10003', '222', '5');
INSERT INTO ZALIK_STUDENT1234(NUMB_FK, SUBJ_ID_FK, MARK)
VALUES('10003', '444', '5')

Reply:
![capture](https://user-images.githubusercontent.com/22836311/38943683-b1ec192c-433a-11e8-9bde-e9fb5dee791e.PNG)

Reply:
INSERT INTO "FURRR"."STUD" (ST_NAME, ST_GROUP, ST_ID, ST_FACULTY) VALUES ('MUR', '61', '1', 'A')
INSERT INTO "FURRR"."STUD" (ST_NAME, ST_GROUP, ST_ID, ST_FACULTY) VALUES ('FUR', '62', '2', 'A')
INSERT INTO "FURRR"."STUD" (ST_NAME, ST_GROUP, ST_ID, ST_FACULTY) VALUES ('KRYA', '51', '3', 'B')
INSERT INTO "FURRR"."STUD" (ST_NAME, ST_GROUP, ST_ID, ST_FACULTY) VALUES ('GAV', '52', '4', 'C')

INSERT INTO "FURRR"."SUBJ" (SBJ_NAME, ST_ID, SBJ_ID) VALUES ('MATHS', '1', '11')
INSERT INTO "FURRR"."SUBJ" (SBJ_NAME, ST_ID, SBJ_ID) VALUES ('DB', '3', '22')
INSERT INTO "FURRR"."SUBJ" (SBJ_NAME, ST_ID, SBJ_ID) VALUES ('OOP', '2', '33') 

INSERT INTO "FURRR"."MARK" (M_ID, M_NAME) VALUES ('111', '100')
INSERT INTO "FURRR"."MARK" (M_ID, M_NAME) VALUES ('222', '90')
INSERT INTO "FURRR"."MARK" (M_ID, M_NAME) VALUES ('333', '80')
INSERT INTO "FURRR"."MARK" (M_ID, M_NAME) VALUES ('444', '70')
INSERT INTO "FURRR"."MARK" (M_ID, M_NAME) VALUES ('555', '60') 

INSERT INTO "FURRR"."ZALIK" (SBJ_ID, M_ID, SEMESTR) VALUES ('22', '111', '4')
INSERT INTO "FURRR"."ZALIK" (SBJ_ID, M_ID, SEMESTR) VALUES ('11', '111', '4')
INSERT INTO "FURRR"."ZALIK" (SBJ_ID, M_ID, SEMESTR) VALUES ('33', '222', '4')

Reply:
![capture](https://user-images.githubusercontent.com/22836311/39275850-76ec424c-48ef-11e8-8a79-10c61692a02f.PNG)
![capture1](https://user-images.githubusercontent.com/22836311/39275851-7710c63a-48ef-11e8-8584-fd27ebd0433f.PNG)

Reply:
import cx_Oracle
connection = cx_Oracle.connect("furrr", "furrr", "Test")
cursor = connection.cursor()
cursor.execute("""
SELECT
    STUD_ST_ID,
    STUD.ST_NAME,
    NVL(SUM(ZALIK.M_ID),0) as MARK_SUM
 FROM
    STUD LEFT JOIN SUBJ
        ON STUD.ST_ID = SUBJ.ST_ID
    LEFT JOIN ZALIK
        ON SUBJ.SBJ_ID = ZALIK.SBJ_ID
    GROUP BY STUD.ST_ID, STUD.ST_NAME) """)
 
students = []
marks_sum = []
 for row in cursor:
    print("Student name: ",row[1]," and his mark sum: ",row[2])
    students += [row[0]+" "+row[1]]
    marks_sum += [row[2]]
 data = [go.Bar(
            x=customers,
            y=orders_sum
    )]
 
layout = go.Layout(
    title='Students and Markss sum',
    xaxis=dict(
        title='Students',
        titlefont=dict(
            family='Courier New, monospace',
            size=18,
            color='#7f7f7f'
        )
    ),
    yaxis=dict(
        title='Subjects sum',
        rangemode='nonnegative',
        autorange=True,
        titlefont=dict(
            family='Courier New, monospace',
            size=18,
            color='#7f7f7f'
        )
    )
)
fig = go.Figure(data=data, layout=layout)
students_marks_sum = py.plot(fig)

Reply:
import cx_Oracle

  import   plotly.offline as py
  import   plotly.graph_objs as go
   
   
  connection   = cx_Oracle.connect("studentpma", "studentpma", "server-pc/xe")
   
  cursor   = connection.cursor()
   
  cursor.execute("""
  SELECT
  STUD.ST_ID),
  STUD.ST_NAME),
  NVL(SUM(ZALIK.M_ID),0)   as M
  FROM
  STUDS   LEFT JOIN SUBJ
  ON   STUD.ST_ID = STUD.ST_ID
  LEFT   JOIN ZALIK
  ON   SUBJ.SBJ_ID = ZALIK.SBJ_ID
  GROUP   BY STUD.ST_ID,STUD.ST_NAME
  """);
   
  STUDENTS   = []
  M =   []
   
   
  for row in cursor:
  print("Student name   ",row[1]+" id: ("+row[0],") and his mark   sum: ",row[2])
  vendors   += [row[1]+" "+row[0]]
  products_sum   += [row[2]]
 
  |  
  | pie =   go.Pie(labels=vendors, values=products_sum)
  | py.plot([pie])

Reply:
import cx_Oracle
--
   import   plotly.offline as py
  import   plotly.graph_objs as go
    
    
   connection   = cx_Oracle.connect("furrr", "furrr", "Test")
    
   cursor   = connection.cursor()
    
    
    
   cursor.execute("""
   SELECT
   SUBJ.SEMESTR,
   NVL(SUM(ZALIK.MARK),0)   as M
   FROM
   SUBJ   LEFT JOIN ZALIK
   ON   SUBJ.SBJ_ID = ZALIK.SBJ_ID
   GROUP   BY SUBJ.SBJ_ID
   ORDER   BY SUBJ.SEMESTR
   """)
   SUBJ_SEMESTR   = []
   M =   []
    
    
   for row in cursor:
   print("Date ",row[0]," sum: ",row[1])
   Subj_semestr   += [row[0]]
   mark   += [row[1]]
    
    
   subject_date_marks   = go.Scatter(
   x=subj_semestr,
   y=mark,
   mode='lines+markers'
   )
   data   = [subject date mark]
   py.plot(data)